File: scripts/storage-comparison.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Offline comparison of four status representations.  It creates disposable
# Git repositories only; no fixture, cache, or status view is retained.
require "fileutils"
require "open3"
require "tmpdir"

NODES = 100
ACTIVE = 25
BASELINE = File.expand_path("../benchmark/storage-comparison-baseline.txt", __dir__)

def run(*command, chdir:)
  out, status = Open3.capture2e(*command, chdir: chdir)
  raise "#{command.join(" ")}: #{out}" unless status.success?
  out
end

def node(id, status = nil)
  metadata = ["---", "context_rev: 1", "updated: 2026-09-10T00:00:00Z", "summary: Fixture #{id}."]
  metadata << "status: #{status}" if status
  (metadata + ["---", "", "Area [[IDX-900-root]].", ""]).join("\n")
end

def git_setup(root)
  run("git", "init", "-q", chdir: root)
  run("git", "config", "user.email", "fixture@example.invalid", chdir: root)
  run("git", "config", "user.name", "Fixture", chdir: root)
end

def commit(root, message)
  run("git", "add", ".", chdir: root)
  run("git", "commit", "-qm", message, chdir: root)
end

def merge_conflicts(root, transition)
  base = run("git", "branch", "--show-current", chdir: root).strip
  run("git", "checkout", "-qb", "left", chdir: root)
  transition.call("TAS-00001")
  commit(root, "left transition")
  run("git", "checkout", "-q", base, chdir: root)
  run("git", "checkout", "-qb", "right", chdir: root)
  transition.call("TAS-00002")
  commit(root, "right transition")
  output, status = Open3.capture2e("git", "merge", "left", chdir: root)
  status.success? ? 0 : output.scan(/^UU /).length + output.scan(/^CONFLICT /).length
end

def diff_shape(root, transition)
  transition.call("TAS-00003")
  run("git", "add", "-A", chdir: root)
  codes = run("git", "diff", "--cached", "--name-status", "-M", chdir: root).lines.map(&:split).map(&:first)
  [codes.sort.join("+"), codes.sum { |code| code.start_with?("R", "C") ? 2 : 1 }]
end

def restore(root)
  run("git", "reset", "--hard", "-q", "HEAD", chdir: root)
  run("git", "clean", "-fdq", chdir: root)
end

def status_count(paths)
  paths.count { |path| File.exist?(path) }
end

def directory_case(root)
  %w[active proposed blocked resolved].each { |status| FileUtils.mkdir_p(File.join(root, "nodes", status)) }
  NODES.times do |i|
    id = format("TAS-%05d", i + 1)
    status = i < ACTIVE ? "active" : "resolved"
    File.write(File.join(root, "nodes", status, "#{id}.md"), node(id))
  end
  query_entries = status_count(Dir.glob(File.join(root, "nodes", "active", "*.md")))
  commit(root, "baseline")
  transition = lambda do |id|
    FileUtils.mv(File.join(root, "nodes", "active", "#{id}.md"), File.join(root, "nodes", "resolved", "#{id}.md"))
  end
  shape, paths = diff_shape(root, transition)
  restore(root)
  conflicts = merge_conflicts(root, transition)
  { shape: shape, paths: paths, query_reads: 0, query_entries: query_entries, conflicts: conflicts, stale: 0, stable: 0 }
end

def stationary_case(root)
  canonical = File.join(root, "nodes", "canonical")
  NODES.times do |i|
    id = format("TAS-%05d", i + 1)
    FileUtils.mkdir_p(File.join(canonical, id[-2, 2]))
    File.write(File.join(canonical, id[-2, 2], "#{id}.md"), node(id, i < ACTIVE ? "active" : "resolved"))
  end
  paths = Dir.glob(File.join(canonical, "*", "*.md"))
  query_entries = paths.count { |path| File.read(path).include?("status: active") }
  commit(root, "baseline")
  transition = lambda do |id|
    path = Dir.glob(File.join(canonical, "*", "#{id}.md")).fetch(0)
    File.write(path, File.read(path).sub("status: active", "status: resolved"))
  end
  shape, paths_changed = diff_shape(root, transition)
  restore(root)
  conflicts = merge_conflicts(root, transition)
  { shape: shape, paths: paths_changed, query_reads: paths.length, query_entries: query_entries, conflicts: conflicts, stale: 0, stable: 1 }
end

def symlink_case(root)
  canonical = File.join(root, "nodes", "canonical")
  %w[active resolved].each { |status| FileUtils.mkdir_p(File.join(root, "nodes", "status", status)) }
  NODES.times do |i|
    id = format("TAS-%05d", i + 1)
    source = File.join(canonical, id[-2, 2], "#{id}.md")
    FileUtils.mkdir_p(File.dirname(source))
    File.write(source, node(id))
    status = i < ACTIVE ? "active" : "resolved"
    File.symlink(File.join("..", "..", "canonical", id[-2, 2], "#{id}.md"), File.join(root, "nodes", "status", status, "#{id}.md"))
  end
  query_entries = status_count(Dir.glob(File.join(root, "nodes", "status", "active", "*.md")))
  commit(root, "baseline")
  transition = lambda do |id|
    old = File.join(root, "nodes", "status", "active", "#{id}.md")
    new = File.join(root, "nodes", "status", "resolved", "#{id}.md")
    FileUtils.mv(old, new)
  end
  shape, paths = diff_shape(root, transition)
  restore(root)
  conflicts = merge_conflicts(root, transition)
  # A deleted view entry is invisible to a directory query although its canonical
  # node still exists; this is the stale/broken-view failure mode.
  FileUtils.rm(File.join(root, "nodes", "status", "active", "TAS-00004.md"))
  stale_link = File.join(root, "nodes", "status", "active", "TAS-00004.md")
  source = Dir.glob(File.join(canonical, "*", "TAS-00004.md")).fetch(0)
  stale = File.exist?(source) && !File.exist?(stale_link) ? 1 : 0
  { shape: shape, paths: paths, query_reads: 0, query_entries: query_entries, conflicts: conflicts, stale: stale, stable: 1 }
end

def index_case(root)
  canonical = File.join(root, "nodes", "canonical")
  NODES.times do |i|
    id = format("TAS-%05d", i + 1)
    FileUtils.mkdir_p(File.join(canonical, id[-2, 2]))
    File.write(File.join(canonical, id[-2, 2], "#{id}.md"), node(id))
  end
  active = (1..ACTIVE).map { |n| format("TAS-%05d", n) }
  index_path = File.join(root, "nodes", "status-index.md")
  write_index = lambda { |ids| File.write(index_path, "active: #{ids.join(",")}\n") }
  read_index = lambda do
    match = /^active: (.*)$/.match(File.read(index_path))
    raise "missing active status index" unless match
    match[1].split(",").reject(&:empty?)
  end
  write_index.call(active)
  query_entries = active.length
  commit(root, "baseline")
  transition = lambda do |id|
    # Read the checked-out fixture state for each scenario.  A closure over a
    # mutable array would let diff-shape mutations leak into the two branches.
    write_index.call(read_index.call - [id])
  end
  shape, paths = diff_shape(root, transition)
  restore(root)
  conflicts = merge_conflicts(root, transition)
  # Removing a name from the cache leaves the canonical file untouched.
  write_index.call(read_index.call - ["TAS-00004"])
  stale = File.exist?(Dir.glob(File.join(canonical, "*", "TAS-00004.md")).fetch(0)) ? 1 : 0
  { shape: shape, paths: paths, query_reads: 1, query_entries: query_entries, conflicts: conflicts, stale: stale, stable: 1 }
end

def result_for(name)
  Dir.mktmpdir("kg-storage-") do |root|
    git_setup(root)
    send("#{name}_case", root)
  end
end

expected = File.readlines(BASELINE, chomp: true).reject { |line| line.empty? || line.start_with?("#") }.to_h { |line| line.split(":", 2) }
%w[directory stationary symlink index].each do |name|
  r = result_for(name)
  line = [r[:shape], r[:paths], r[:query_reads], r[:query_entries], r[:conflicts], r[:stale], r[:stable]].join(",")
  abort "baseline mismatch for #{name}: expected #{expected.fetch(name)}, got #{line}" if ARGV == ["--verify"] && expected.fetch(name) != line
  puts "storage{name,diff,transition_paths,query_reads,query_entries,merge_conflicts,stale_view,stable_canonical_path}: #{name},#{line}"
end
puts "verification: passed" if ARGV == ["--verify"]

File: src/braintree/storage_comparison.py
"""Offline comparison of four status representations.

Typed Python port of ``scripts/storage-comparison.rb``. It creates disposable
Git repositories only; no fixture, cache, or status view is retained. The
``storage{...}`` line and the tracked baseline in
``benchmark/storage-comparison-baseline.txt`` are preserved.
"""

from __future__ import annotations

import glob
import os
import re
import subprocess
import sys
import tempfile
from collections.abc import Callable, Sequence
from dataclasses import dataclass
from pathlib import Path

__all__ = ["main"]

NODES = 100
ACTIVE = 25
_UU = re.compile(r"^UU ", re.MULTILINE)
_CONFLICT = re.compile(r"^CONFLICT ", re.MULTILINE)
_ACTIVE_INDEX = re.compile(r"^active: (.*)$", re.MULTILINE)

Transition = Callable[[str], None]


class _CaseError(Exception):
    """A Git fixture command failed."""


@dataclass
class _CaseResult:
    shape: str
    transition_paths: int
    query_reads: int
    query_entries: int
    conflicts: int
    stale: int
    stable: int


def _repo_root() -> Path:
    return Path(__file__).resolve().parents[2]


def _baseline_path() -> Path:
    return _repo_root() / "benchmark" / "storage-comparison-baseline.txt"


def _capture(*command: str, chdir: str) -> tuple[str, bool]:
    result = subprocess.run(
        command,
        cwd=chdir,
        stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT,
        text=True,
    )
    return result.stdout or "", result.returncode == 0


def _run(*command: str, chdir: str) -> str:
    output, success = _capture(*command, chdir=chdir)
    if not success:
        raise _CaseError(f"{' '.join(command)}: {output}")
    return output


def _node(node_id: str, status: str | None = None) -> str:
    metadata = [
        "---",
        "context_rev: 1",
        "updated: 2026-09-10T00:00:00Z",
        f"summary: Fixture {node_id}.",
    ]
    if status is not None:
        metadata.append(f"status: {status}")
    metadata.extend(["---", "", "Area [[IDX-900-root]].", ""])
    return "\n".join(metadata)


def _git_setup(root: str) -> None:
    _run("git", "init", "-q", chdir=root)
    _run("git", "config", "user.email", "fixture@example.invalid", chdir=root)
    _run("git", "config", "user.name", "Fixture", chdir=root)


def _commit(root: str, message: str) -> None:
    _run("git", "add", ".", chdir=root)
    _run("git", "commit", "-qm", message, chdir=root)


def _merge_conflicts(root: str, transition: Transition) -> int:
    base = _run("git", "branch", "--show-current", chdir=root).strip()
    _run("git", "checkout", "-qb", "left", chdir=root)
    transition("TAS-00001")
    _commit(root, "left transition")
    _run("git", "checkout", "-q", base, chdir=root)
    _run("git", "checkout", "-qb", "right", chdir=root)
    transition("TAS-00002")
    _commit(root, "right transition")
    output, success = _capture("git", "merge", "left", chdir=root)
    if success:
        return 0
    return len(_UU.findall(output)) + len(_CONFLICT.findall(output))


def _diff_shape(root: str, transition: Transition) -> tuple[str, int]:
    transition("TAS-00003")
    _run("git", "add", "-A", chdir=root)
    output = _run("git", "diff", "--cached", "--name-status", "-M", chdir=root)
    codes = [line.split()[0] for line in output.splitlines() if line.split()]
    shape = "+".join(sorted(codes))
    weight = sum(2 if code.startswith(("R", "C")) else 1 for code in codes)
    return shape, weight


def _restore(root: str) -> None:
    _run("git", "reset", "--hard", "-q", "HEAD", chdir=root)
    _run("git", "clean", "-fdq", chdir=root)


def _status_count(paths: Sequence[str]) -> int:
    return sum(1 for path in paths if os.path.exists(path))


def _directory_case(root: str) -> _CaseResult:
    for status in ("active", "proposed", "blocked", "resolved"):
        os.makedirs(os.path.join(root, "nodes", status), exist_ok=True)
    for index in range(NODES):
        node_id = f"TAS-{index + 1:05d}"
        status = "active" if index < ACTIVE else "resolved"
        path = os.path.join(root, "nodes", status, f"{node_id}.md")
        with open(path, "w", encoding="utf-8") as handle:
            handle.write(_node(node_id))
    query_entries = _status_count(glob.glob(os.path.join(root, "nodes", "active", "*.md")))
    _commit(root, "baseline")

    def transition(node_id: str) -> None:
        os.replace(
            os.path.join(root, "nodes", "active", f"{node_id}.md"),
            os.path.join(root, "nodes", "resolved", f"{node_id}.md"),
        )

    shape, paths = _diff_shape(root, transition)
    _restore(root)
    conflicts = _merge_conflicts(root, transition)
    return _CaseResult(shape, paths, 0, query_entries, conflicts, 0, 0)


def _stationary_case(root: str) -> _CaseResult:
    canonical = os.path.join(root, "nodes", "canonical")
    for index in range(NODES):
        node_id = f"TAS-{index + 1:05d}"
        directory = os.path.join(canonical, node_id[-2:])
        os.makedirs(directory, exist_ok=True)
        with open(os.path.join(directory, f"{node_id}.md"), "w", encoding="utf-8") as handle:
            handle.write(_node(node_id, "active" if index < ACTIVE else "resolved"))
    paths = glob.glob(os.path.join(canonical, "*", "*.md"))
    query_entries = 0
    for path in paths:
        with open(path, encoding="utf-8") as handle:
            if "status: active" in handle.read():
                query_entries += 1
    _commit(root, "baseline")

    def transition(node_id: str) -> None:
        path = glob.glob(os.path.join(canonical, "*", f"{node_id}.md"))[0]
        with open(path, encoding="utf-8") as handle:
            text = handle.read()
        with open(path, "w", encoding="utf-8") as handle:
            handle.write(text.replace("status: active", "status: resolved"))

    shape, paths_changed = _diff_shape(root, transition)
    _restore(root)
    conflicts = _merge_conflicts(root, transition)
    return _CaseResult(shape, paths_changed, len(paths), query_entries, conflicts, 0, 1)


def _symlink_case(root: str) -> _CaseResult:
    canonical = os.path.join(root, "nodes", "canonical")
    for status in ("active", "resolved"):
        os.makedirs(os.path.join(root, "nodes", "status", status), exist_ok=True)
    for index in range(NODES):
        node_id = f"TAS-{index + 1:05d}"
        directory = os.path.join(canonical, node_id[-2:])
        os.makedirs(directory, exist_ok=True)
        source = os.path.join(directory, f"{node_id}.md")
        with open(source, "w", encoding="utf-8") as handle:
            handle.write(_node(node_id))
        status = "active" if index < ACTIVE else "resolved"
        link = os.path.join(root, "nodes", "status", status, f"{node_id}.md")
        os.symlink(os.path.join("..", "..", "canonical", node_id[-2:], f"{node_id}.md"), link)
    query_entries = _status_count(
        glob.glob(os.path.join(root, "nodes", "status", "active", "*.md"))
    )
    _commit(root, "baseline")

    def transition(node_id: str) -> None:
        old = os.path.join(root, "nodes", "status", "active", f"{node_id}.md")
        new = os.path.join(root, "nodes", "status", "resolved", f"{node_id}.md")
        os.replace(old, new)

    shape, paths = _diff_shape(root, transition)
    _restore(root)
    conflicts = _merge_conflicts(root, transition)
    # A deleted view entry is invisible to a directory query although its
    # canonical node still exists; this is the stale/broken-view failure mode.
    link = os.path.join(root, "nodes", "status", "active", "TAS-00004.md")
    os.remove(link)
    source = glob.glob(os.path.join(canonical, "*", "TAS-00004.md"))[0]
    stale = 1 if os.path.exists(source) and not os.path.exists(link) else 0
    return _CaseResult(shape, paths, 0, query_entries, conflicts, stale, 1)


def _index_case(root: str) -> _CaseResult:
    canonical = os.path.join(root, "nodes", "canonical")
    for index in range(NODES):
        node_id = f"TAS-{index + 1:05d}"
        directory = os.path.join(canonical, node_id[-2:])
        os.makedirs(directory, exist_ok=True)
        with open(os.path.join(directory, f"{node_id}.md"), "w", encoding="utf-8") as handle:
            handle.write(_node(node_id))
    active = [f"TAS-{number:05d}" for number in range(1, ACTIVE + 1)]
    index_path = os.path.join(root, "nodes", "status-index.md")

    def write_index(ids: Sequence[str]) -> None:
        with open(index_path, "w", encoding="utf-8") as handle:
            handle.write(f"active: {','.join(ids)}\n")

    def read_index() -> list[str]:
        with open(index_path, encoding="utf-8") as handle:
            match = _ACTIVE_INDEX.search(handle.read())
        if match is None:
            raise _CaseError("missing active status index")
        return [value for value in match.group(1).split(",") if value]

    write_index(active)
    query_entries = len(active)
    _commit(root, "baseline")

    def transition(node_id: str) -> None:
        # Read the checked-out fixture state for each scenario. A closure over
        # a mutable list would let diff-shape mutations leak into the branches.
        write_index([value for value in read_index() if value != node_id])

    shape, paths = _diff_shape(root, transition)
    _restore(root)
    conflicts = _merge_conflicts(root, transition)
    # Removing a name from the cache leaves the canonical file untouched.
    write_index([value for value in read_index() if value != "TAS-00004"])
    source = glob.glob(os.path.join(canonical, "*", "TAS-00004.md"))[0]
    stale = 1 if os.path.exists(source) else 0
    return _CaseResult(shape, paths, 1, query_entries, conflicts, stale, 1)


_CASES: dict[str, Callable[[str], _CaseResult]] = {
    "directory": _directory_case,
    "stationary": _stationary_case,
    "symlink": _symlink_case,
    "index": _index_case,
}


def _result_for(name: str) -> _CaseResult:
    with tempfile.TemporaryDirectory(prefix="bt-storage-") as root:
        _git_setup(root)
        return _CASES[name](root)


def _baseline() -> dict[str, str]:
    expected: dict[str, str] = {}
    for line in _baseline_path().read_text(encoding="utf-8").splitlines():
        if not line or line.startswith("#"):
            continue
        name, value = line.split(":", 2)
        expected[name] = value
    return expected


def main(argv: Sequence[str] | None = None) -> int:
    """Run the storage-representation comparison and return the exit code."""
    args = list(sys.argv[1:] if argv is None else argv)
    expected = _baseline()
    for name in _CASES:
        try:
            result = _result_for(name)
        except _CaseError as error:
            print(str(error), file=sys.stderr)
            return 1
        line = ",".join(
            str(value)
            for value in (
                result.shape,
                result.transition_paths,
                result.query_reads,
                result.query_entries,
                result.conflicts,
                result.stale,
                result.stable,
            )
        )
        if args == ["--verify"] and expected[name] != line:
            print(
                f"baseline mismatch for {name}: expected {expected[name]}, got {line}",
                file=sys.stderr,
            )
            return 1
        print(
            "storage{name,diff,transition_paths,query_reads,query_entries,"
            f"merge_conflicts,stale_view,stable_canonical_path}}: {name},{line}"
        )
    if args == ["--verify"]:
        print("verification: passed")
    return 0


if __name__ == "__main__":  # pragma: no cover - module entry point
    raise SystemExit(main())
